Homework 9

JeRusakow opened the pull request.

JeRusakow committed:
Initial commit for the nineth task

RomanTsaregorodtsev commented on homework9/task1/hw1.py:
```diff
@@ -0,0 +1,57 @@
+    def integer_from_file(file: str) -> Generator[int, None, None]:
+        """Creates int generator for the given file"""
+        with open(file, "r") as f:
+            yield from (int(line) for line in f)
```
1. `f` - это плохое название. Можешь использовать спокойно file. Это не зарезервированное слово в питоне.
2. `f` - это уже итератор, поэтому можно сразу с него делать `yield from`

RomanTsaregorodtsev commented on homework9/task1/hw1.py:
```diff
@@ -0,0 +1,57 @@
+            yield from (int(line) for line in f)
+        yield None
+
+    def find_min(numbers: List[Union[int, None]]) -> Union[int, None]:
+        """A modified min(), because original min() does not support NoneType"""
+        minimum = None
+
+        for num in numbers:
+            if minimum is None or (num is not None and num < minimum):
+                minimum = num
+
+        return minimum
```
Вообще я могу упростить твой хак просто воспользовавшись `float("inf")` в строке 37. Тогда тебе не придётся городить кастомную функцию минимума.

RomanTsaregorodtsev commented on homework9/task1/hw1.py:
```diff
@@ -0,0 +1,57 @@
+    num_generators = [integer_from_file(file) for file in file_list]
+    integers = [next(gen) for gen in num_generators]
+
+    while (min_num := find_min(integers)) is not None:
```
В общем решение рабочее, но всё равно кажется, как будто можно проще. Тебя не смущает, что на каждом шаге тебе приходится искать минимальное значение? Может можно использовать какую-то структуру данных, которая упростила бы этот процесс? За дискуссией милости прошу в телеграм

RomanTsaregorodtsev commented on homework9/task2/hw2.py:
```diff
@@ -0,0 +1,52 @@
+        None
+    """
+
+    def __init__(self, exception_type: Type[BaseException]):
+        self.exception_type = exception_type
+
```
Теперь давай сделаем, чтобы можно было принимать любое количество ошибок через запятую

RomanTsaregorodtsev commented on homework9/task2/hw2.py:
```diff
@@ -0,0 +1,52 @@
+
+
+@contextmanager
+def suppress_exception(exception_type: Type[BaseException]) -> None:
```
Здесь аналогично

RomanTsaregorodtsev commented on homework9/task3/hw3.py:
```diff
@@ -0,0 +1,54 @@
+    if not os.path.exists(dir_path):
+        raise FileNotFoundError(f"Path '{dir_path}' does not exists")
+
+    file_list = os.listdir(dir_path)
```
1. Рекомендую при работе с файлами использовать модуль `pathlib`. Он позволяет работать с путём как с объектом и там есть много удобных методов
2. Давай распространим условие, чтобы функция принимал флаг `recursive` в зависимости от которого мы бы обходили все вложенные директории или нет. В это тоже поможет модуль `pathlib` так как у объекта есть удобные способо обходить директорию. В общем крайне советую ознакомиться.

RomanTsaregorodtsev commented on tests/test_homework9/test_task3.py:
```diff
@@ -0,0 +1,69 @@
+import random
```
В этих тестах тоже наверное правильно было бы сделать тестовую директорию с файлами

JeRusakow committed:
HW9 first amendment

- Task 1 and task 3 as well as their tests rewritten to use pathlib Path objects
- Exception suppressors from Task 2 now are able to take several exception types
- Test rewritten to use persistent files instead of dynamically generated ones

JeRusakow committed:
Task1 made to be more algorythmically effective

JeRusakow committed:
Fixet empty_directory test failure

JeRusakow committed:
Task1 redone with the use of heapq.merge
